crypto/evp/p_legacy.c

```diff
@@ -26,8 +26,9 @@ int EVP_PKEY_set1_RSA(EVP_PKEY *pkey, RSA *key)
 {
     int ret = EVP_PKEY_assign_RSA(pkey, key);
 
-    if (ret)
-        RSA_up_ref(key);
+    if (ret && !RSA_up_ref(key))
+        return 0;
```

Comment:
Hmm... the key is already assigned. We should up-ref first, then assign if that passes.